### PR TITLE
Task/51372 redirect uri flow

### PR DESCRIPTION
**Pull request by apfohl**
https://community.openproject.org/work_packages/51372/activity

This PR moves the "redirect URI" form element from the OAuth section to a separate row. For OneDrive/Sharepoint.

https://www.figma.com/file/gtLQfPe09X7XugAH8L7dTy/File-Storages-Integration?type=design&node-id=6648-77574&mode=design&t=1ZLvTmKG3QG69qGJ-0

### CODE REVIEW

File: modules/storages/app/views/storages/admin/oauth_clients/update.turbo_stream.erb [hunk 1]
```diff
@@ -30,3 +30,7 @@ See COPYRIGHT and LICENSE files for more details.
 <%= turbo_stream.update :storage_oauth_client_section do %>
   <%= render(Storages::Admin::OAuthClientInfoComponent.new(oauth_client: @oauth_client, storage: @storage)) %>
 <% end %>
+
+<%= turbo_stream.update :storage_redirect_url_section do %>
```
Comment (Kharonus) [inline modules/storages/app/views/storages/admin/oauth_clients/update.turbo_stream.erb]: 🟡 minor: use `:storage_redirect_uri_section`

Comment (Kharonus) [inline modules/storages/app/views/storages/admin/oauth_clients/update.turbo_stream.erb]: 🟡 minor: use `:storage_redirect_uri_section`

File: modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb [hunk 2]
```diff
@@ -66,21 +66,27 @@ def create # rubocop:disable Metrics/AbcSize
...
-      else
-        raise "Unsupported provider type: #{@storage.short_provider_type}"
+        #   respond_to do |format|
+        #     format.turbo_stream { render :create }
+        #   end
+        # elsif @storage.provider_type_one_drive?
+        #   flash[:primer_banner] = { message: I18n.t(:'storages.notice_successful_storage_connection'), scheme: :success }
+        #   redirect_to admin_settings_storages_path
+        # else
+        #   raise "Unsupported provider type: #{@storage.short_provider_type}"
```
Comment (Kharonus) [inline modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: 🟡 this is commented, so we can remove it? you know, were the success flash for onedrive is raised instead?

Comment (Kharonus) [reply modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: I guess it was moved to the `finish_setup`? And this is called from primer form destination URL?

Comment (Kharonus) [reply modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: A question related to that, also to @akabiru :

Why does this flash not exist for nextcloud? nc form does not close upon submitting last step.

**Review by Kharonus — COMMENTED** [review]
I have no really big issues, but I think, they are worth discussing before merging.

File: modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb (hunk 2 above)
Comment (akabiru) [reply modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: > A question related to that, also to @akabiru :
> 
> Why does this flash not exist for nextcloud? nc form does not close upon submitting last step.

For Nextcloud, the Final Step is automatic management, so the flash is set there: https://github.com/opf/openproject/blob/dev/modules/storages/app/controllers/storages/admin/automatically_managed_project_folders_controller.rb#L77-L81

Comment (Kharonus) [reply modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: But it does not happen if I click "Complete without it", at least not in my local instance on branch `release/13.1`. Is this intended or a bug?

Comment (apfohl) [reply modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: > I guess it was moved to the `finish_setup`? And this is called from primer form destination URL?

Yes that is right. Technical explanation is difficult. But I can provide if needed.

File: modules/storages/app/components/storages/admin/forms/oauth_client_form_component.html.erb [hunk 3]
```diff
@@ -27,8 +27,8 @@
               client_id_input_options: {
                 data: {
                   'storages--oauth-client-form-target': 'clientId',
-                  action: 'input->storages--oauth-client-form#toggleSubmitButtonDisabled '\
-                          'input->storages--oauth-client-form#setRedirectUriValue'
+                  action: 'input->storages--oauth-client-form#toggleSubmitButtonDisabled ' \
+                    'input->storages--oauth-client-form#setRedirectUriValue'
```
Comment (akabiru) [inline modules/storages/app/components/storages/admin/forms/oauth_client_form_component.html.erb:31]: 🟠  🧹 We no longer need the prefill actions- I've cleaned it up https://github.com/opf/openproject/pull/14366

File: modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb (hunk 2 above)
Comment (akabiru) [reply modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: > But it does not happen if I click "Complete without it", at least not in my local instance on branch `release/13.1`. Is this intended or a bug?

The flash is only rendered on first time configuration- it does not render on edit. Let's pair if you're seeing alternate behavior. My local tests confirm this.

Comment (akabiru) [reply modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: > 🟡 this is commented, so we can remove it? you know, were the success flash for onedrive is raised instead?

The commented out code should be deleted, we only need to `prepare_storage_for_automatic_management_form` for nextcloud storage

**Review by akabiru — COMMENTED** [review]
Coming along nicely, just some cleanup comments- I tried to assist 😄 🧹 

https://github.com/opf/openproject/pull/14366

File: modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb (hunk 2 above)
Comment (apfohl) [reply modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: @akabiru Okay, so only thing left is, hide the flash on edit?

Comment (Kharonus) [reply modules/storages/app/controllers/storages/admin/oauth_clients_controller.rb]: we found a proper solution: on viewing the redirect URI in a complete setup (editing a storage), the submit button is no longer enabled, as there is no "setup to complete". this leads to not having a redirect to the list and no flash anymore.

File: modules/storages/app/components/storages/admin/forms/oauth_client_form_component.html.erb (hunk 3 above)
Comment (akabiru) [reply modules/storages/app/components/storages/admin/forms/oauth_client_form_component.html.erb:31]: This is unresolved, I'll update the PR and merge on green https://github.com/opf/openproject/pull/14366

**Review by akabiru — APPROVED** [review]
LGTM 👍🏾 

----

I would still prefer to have the dead code cleaned up (I raised https://github.com/opf/openproject/pull/14366) it's fine if you want to merge first, I can followup separately and merge https://github.com/opf/openproject/pull/14366 once tests are green.

File: modules/storages/app/components/storages/admin/forms/oauth_client_form_component.html.erb (hunk 3 above)
Comment (akabiru) [reply modules/storages/app/components/storages/admin/forms/oauth_client_form_component.html.erb:31]: Resolved

**Pull request merged**